pyro/ops/welford.py

```diff
@@ -0,0 +1,70 @@
+from collections import defaultdict
+
+import torch
+
+
+class WelfordCovariance(object):
+    """
+    Implements Welford's online scheme for estimating (co)variance (see :math:`[1]`).
+    Useful for adapting diagonal and dense mass structures for HMC.
+
+    **References**
+
+    [1] `The Art of Computer Programming`,
+    Donald E. Knuth
+    """
+    def __init__(self, diagonal=True):
+        self.diagonal = diagonal
+        self.reset()
+
+    def _deltas(self, sample):
+        deltas = []
+        for i, x in enumerate(sample):
+            if i > len(self.means) - 1:
+                self.means.append(0.)
+            mean = self.means[i]
+            delta_pre = x - mean
+            updated_mean = mean + delta_pre / self.n_samples
+            delta_post = x - updated_mean
+            deltas.append((updated_mean, delta_pre, delta_post))
+        return deltas
+
+    def _unroll(self, sample):
+        unrolled = []
+        for rolled in sample:
+            unrolled.extend(rolled.reshape(-1).data.tolist())
```

Comment:
make `unrolled` a list of "unrolled" tensors or a unrolled+joined tensor, rather than "unroll+joined" list of floats.